File: leetcode/0442-find-all-duplicates-in-an-array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDuplicates = function (nums) {
  let numsCount = {};
  let res = [];

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] in numsCount) numsCount[nums[i]]++;
    else numsCount[nums[i]] = 1;
  }

  for (num in numsCount) {
    if (numsCount[num] > 1) res.push(num);
  }

  return res;
};

File: leetcode/0442-find-all-duplicates-in-an-array.py
# def findDuplicates(nums):
#     counts = {}
#     results = []

#     for num in nums:
#         if num in counts:
#             counts[num] += 1
#         else:
#             counts[num] = 1

#     for k, v in counts.items():
#         if v > 1:
#             results.append(k)

#     return results


def findDuplicates(nums):
    nums_set = set()
    dupes = set()

    for num in nums:
        if num not in nums_set:
            nums_set.add(num)
        else:
            dupes.add(num)

    return list(dupes)
